jax/experimental/jax2tf/jax_export.py

```diff
@@ -215,9 +169,147 @@ def serialize_native(fun_jax: Callable,
       mlir_module=mlir_module,
       mlir_module_serialized=mlir_module_serialized,
       module_kept_var_idx=module_kept_var_idx,
-      xla_call_module_version=xla_call_module_version,
-      dim_args_spec=dim_args_spec
-  )
+      xla_call_module_version=xla_call_module_version)
+
+
+def compute_dim_vars(module: mlir.ir.Module,
+                     args_avals: Sequence[core.ShapedArray]) -> mlir.ir.Module:
+  """Wraps the lowered module with a new "main" that computes the dim vars.
+
+  JAX lowering in presence of shape polymorphism produces a `module` that
+  takes one or more dimension arguments, of type i32 or i64, followed
+  by the regular array arguments. The dimension arguments correspond to the
+  dimension variables appearing in the `args_avals`, in sorted order.
+
+  Consider the lowering of a function with one array argument of type "f32[w, h]",
+  where "w" and "h" are two dimension variables. The `module` will also
+  contain two dimension arguments, corresponding to "h" and "w" respectively::
+
+      func public main(arg_h: i32, arg_w: i32, arg: f32[?, ?]) {
+        ...
+      }
+
+      we rename "main" to "_wrapped_jax_export_main" and add a new "main":
+
+      func public main(arg: f32[?, ?]) {
+         arg_h = hlo.get_dimension_size(arg, 1)
+         arg_w = hlo.get_dimension_size(arg, 0)
+         res = call _wrapped_jax_export_main(arg_h, arg_w, arg)
+         return res
+      }
+
+  Args:
+    module: the HLO module as obtained from lowering. May have a number of
+      dimension arguments, followed by the kept array arguments.
+    args_avals: the avals for all the arguments of the lowered function, which
+      correspond to the array arguments of the `module`.
+
+  Returns the wrapped module.
+  """
+  dim_args_builders = get_dim_arg_builders(args_avals)
+
+  # Make a new module, do not mutate the "module" because it may be cached
+  context = mlir.make_ir_context()
+  with context, ir.Location.unknown(context):
+    # TODO: is there a better way to make a copy of the `module`?
+    new_module = ir.Module.parse(str(module))
```

Comment:
> TODO: is there a better way to make a copy of the `module`?

I've seen the question of moving a module between Contexts's a few times lately - the consensus seems to be that re-parsing is the best option. Better to use bytecode instead of string though. I believe it would be: `ir.Module.parse(mlir.module_to_bytecode(module))`